components/script/dom/htmlimageelement.rs

```diff
@@ -789,26 +791,38 @@ impl HTMLImageElement {
         // Step 2 abort if user-agent does not supports images
         // NOTE: Servo only supports images, skipping this step
 
-        // step 3, 4
-        // TODO: take srcset and parent images into account
-        if !src.is_empty() {
-            // TODO: take pixel density into account
+        // Step 3, 4
+        let mut selected_source = None;
+        let mut pixel_density = None;
+        let src_set = elem.get_string_attribute(&local_name!("srcset"));
+        let is_parent_picture = elem.upcast::<Node>().GetParentElement().unwrap().is::<HTMLPictureElement>();
```

Comment:
Unwrapping here will panic for any element that has no parent. We should use `.map_or(false, |p| p.is::<HTMLPictureElement>())` instead.